app/src/CoffeeShopApp.kt

```diff
@@ -0,0 +1,33 @@
+import java.io.File
+import java.io.FileNotFoundException
+
+class CoffeeShopApp (val x: Double, val y: Double, val filename: String) {
+    fun getNearestShops(): List<CoffeeShopWithDistance> {
+        val user = Location(x, y)
+        return readAllLinesAsCsvRows(filename).toCoffeeShops().toCoffeeShopsWithDistance(user).sortedByDistance().take(3)
+    }
+
+    fun getOutput(): String {
+        try {
+            val shops = getNearestShops()
+
+            if (shops.isEmpty()) {
+                return "No coffee shops was found in $filename."
+            }
+            else {
+                return shops.joinToString(separator = System.lineSeparator())
+            }
+
+        }
+        catch (e: FileNotFoundException) {
+            return "$filename was not found."
+        }
+        catch (e: CouldNotParseCsvRowException) {
```

Comment:
Can you think of some other ways to have resilience but have a nicer user experience?

Reply:
I could ignore the bad rows from the csv file. The user could still use the app with the other available coffee shops.

Reply:
Yeah, good point, also should consider to report the bad rows (all of them) I always find it annoying when apps keep feeding me input errors one by one.